xen/xenops_interface.ml

```diff
@@ -546,11 +548,27 @@ module XenopsAPI (R : RPC) = struct
         Host.guest_agent_feature_list
 
     type cpu_features_array = int64 array [@@deriving rpcty]
+    type policy_array = string [@@deriving rpcty]
+
+    let policy_p =
+      Param.mk ~description:["A cpu policy"]
+        ~name:"policy" policy_array
 
     let cpu_features_array_p =
       Param.mk ~description:["An array containing the raw CPU feature flags"]
         ~name:"features_array" cpu_features_array
 
+    type policy_message = (string * bool * string option) [@@deriving rpcty]
```

Comment:
What is the semantics for the individual tuple members? Would it be better to use a record instead?

Reply:
I looked at the use case:

```
let (sub, eq, err) = Client.HOST.policy_is_compatible dbg left right
```

This suggests to me that a compatibility value needs to be better defined. If we are comparing two sets, this can happen: (1) they are equal, (2) A  is a strict subset of B, (3) B is a strict subset of A, (4) they are incomparable (no subset relation). There are several ways to model this: using two booleans, or a sum type with 4 value.

Reply:
I think a record is a better idea.
I don't disagree that this can be better defined, but I don't think we're dealing with sets anymore. Previously we parsed and dealt with the bitmaps from the output of `xen-cpuid` on a host. These mapped well to sets, and so the operations are mathematical. Now I believe we will be using the policy output from something like `xen-cpuid -p`. This outputs all of the cpu leaves, and msr leaves. And I think that the data represented in this is far more than just 'if it's there it's on', and so I don't think we can do the same level of operations with the information that we hold. Additionally, because we're being more truthful about the information that we lift from hardware, I think the goal is to be more truthful with the errors too - ie instead of just saying that two policies are/aren't equal, to instead lift up a message about why those two policies differ

Perhaps I can tidy up the types that I have defined to make them easier to digest and use, whilst maintaining what I think the Xen team want to be able to achieve with these

Reply:
I am not familiar with `xen-cpuid -p`, it looks like the output below. I think there is a danger of creating something too simple. Maybe it's too late already, but it would be better to create an abstract type (module) with access function. The problem with that is that it cannot be shared across xcp-idl, so we would need to have a representation that can be serialised.

```
[root@lcy2-dt110 ~]# xen-cpuid -p
Xen reports there are maximum 115 leaves and 2 MSRs
Raw policy: 38 leaves, 1 MSRs
 CPUID:
  leaf     subleaf  -> eax      ebx      ecx      edx     
  00000000:ffffffff -> 00000016:756e6547:6c65746e:49656e69
  00000001:ffffffff -> 000906e9:00100800:7ffafbff:bfebfbff
  00000002:ffffffff -> 76036301:00f0b5ff:00000000:00c30000
  00000004:00000000 -> 1c004121:01c0003f:0000003f:00000000
  00000004:00000001 -> 1c004122:01c0003f:0000003f:00000000
  00000004:00000002 -> 1c004143:00c0003f:000003ff:00000000
  00000004:00000003 -> 1c03c163:03c0003f:00001fff:00000006
  00000005:ffffffff -> 00000040:00000040:00000003:00142120
  00000006:ffffffff -> 000027f7:00000002:00000009:00000000
  00000007:00000000 -> 00000000:029c6fbf:00000000:9c002400
  0000000a:ffffffff -> 07300404:00000000:00000000:00000603
  0000000b:00000000 -> 00000001:00000002:00000100:00000000
  0000000b:00000001 -> 00000004:00000008:00000201:00000000
  0000000d:00000000 -> 0000001f:00000440:00000440:00000000
  0000000d:00000001 -> 0000000f:000003c0:00000100:00000000
  0000000d:00000002 -> 00000100:00000240:00000000:00000000
  0000000d:00000003 -> 00000040:000003c0:00000000:00000000
  0000000d:00000004 -> 00000040:00000400:00000000:00000000
  0000000d:00000008 -> 00000080:00000000:00000001:00000000
  80000000:ffffffff -> 80000008:00000000:00000000:00000000
  80000001:ffffffff -> 00000000:00000000:00000121:2c100800
  80000002:ffffffff -> 65746e49:2952286c:6f655820:2952286e
  80000003:ffffffff -> 55504320:2d334520:30333231:20367620
  80000004:ffffffff -> 2e332040:48473035:0000007a:00000000
  80000006:ffffffff -> 00000000:00000000:01006040:00000000
  80000007:ffffffff -> 00000000:00000000:00000000:00000100
  80000008:ffffffff -> 00003027:00000000:00000000:00000000
 MSRs:
  index    -> value           
  000000ce -> 0000000080000000
Host policy: 33 leaves, 1 MSRs
...
```

Reply:
I think the idea was for this format to be a black box for the toolstack, only xen knows about these featuresets and how to compare them.

Not sure how this played with displaying the information via CLI / xenapi